docs/jep/11859-type-annotations.md

```diff
@@ -0,0 +1,160 @@
+```python
+def shuffle(key: KeyArray, x: Array, axis: int = 0) -> jnp.ndarray:
+  ...
+```
+In this case `jnp.ndarray` is an abstract base class that forward-declares the attributes and methods of JAX arrays ([see source](http://google3/third_party/py/jax/_src/numpy/ndarray.py;l=40;rcl=431999528)), and so PyLance in VSCode can offer the full set of autocompletions on results from this function. Here is the result in VSCode:
+
+![VSCode Intellisense Screenshot](../_static/vscode-completion.png)
+
+Listed are all methods and attributes declared by the abstract class
+We'll discuss further below why it was necessary to create this abstract class rather than annotating with `DeviceArray` directly.
+
+### Level 3: Annotations for Static Type-Checking
+
+Static type-checking often is the first thing people think of when considering the purpose of type annotations in Python code.
+While Python does not do any runtime checking of types, several mature static type checking tools exist that can do this as part of a CI test suite.
+The most important ones for JAX are the following:
+
+- [python/mypy](https://github.com/python/mypy) is more or less the standard in the open Python world. JAX currently runs mypy on a subset of source files within the Github Actions CI checks.
+- [google/pytype](https://github.com/google/pytype) is Google's static type checker, and projects which depend on JAX within Google frequently use this.
+- [microsoft/pyright](https://github.com/microsoft/pyright) is important as the static type checker used within VSCode for the PyLance completions mentioned previously.
+
+Full static type checking is the strictest of all the type annotation applications, because it will surface an error any time your type annotations are not precisely correct.
+On the one hand, this is nice because your static type analysis may catch faulty type annotations (for example, a case where a `DeviceArray` method is missing from the `jnp.ndarray` abstract class).
+
+On the other hand, this strictness can make the type checking process very brittle in packages that often rely on duck-typing rather than strict type-safe APIs.
+You'll currently find code comments like `#type: ignore` (for mypy) or `#pytype: disable` (for pytype) peppered throughout the JAX codebase in several hundred places.
+These typically represent cases where typing problems have arisen; they may be inaccuracies in JAX type annotations, or inaccuracies in the static type checker's ability to correctly follow the control flow in the code.
+On occasion, they are due to real & subtle bugs in the behavior of pytype and/or mypy.
+In rare cases, they may be due to the fact that JAX uses Python patterns that are difficult or even impossible to express in terms of Python's static type annotation syntax.
+
+## Type Annotation Challenges for JAX
```

Comment:
Nit: May be best to use sentence case. Because you use lower-case for `pytype` and `mype`, capitalizing some words may look strange. Applies here and across the entire doc. Ref: https://developers.google.com/style/capitalization?hl=en#capitalization-in-titles-and-headings

Example:

```suggestion
## Type annotation challenges for JAX
```

Reply:
@jakevdp Forgot to mention

> "... capitalize only the first word in the title, the first word in a subheading after a colon, and any proper nouns or other terms that are always capitalized a certain way."

From https://developers.google.com/style/capitalization#capitalization-in-titles-and-headings

So:

```diff
- ## Level 1: type annotation is cool
+ ## Level 1: Type annotation is cool
```